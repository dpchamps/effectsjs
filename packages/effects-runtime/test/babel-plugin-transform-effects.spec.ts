import transformEffects from "babel-plugin-effects";
import pluginTester from "babel-plugin-tester";

const fixturesDirectory = `__fixtures__/transformEffects`;

/*
TODO: when the specification is settled, these should be corrected and enabled
 */
describe("transformEffects", () => {
  pluginTester({
    plugin: transformEffects,
    pluginName: "Transform Effects",
    title: "Transform proposed effects keywords into working JS",
    filename: __filename,
    tests: [
      // {
      //   title: "Perform effect",
      //   fixture: `${fixturesDirectory}/perform/perform-simple-effect-in.js`,
      //   outputFixture: `${fixturesDirectory}/perform/perform-simple-effect-out.js`
      // },
      // {
      //   title: "Perform effect with assignment",
      //   fixture: `${fixturesDirectory}/perform/perform-effect-with-assignment-in.js`,
      //   outputFixture: `${fixturesDirectory}/perform/perform-effect-with-assignment-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall Base Case",
      //   fixture: `${fixturesDirectory}/try-handle-recall/basic-in.js`,
      //   outputFixture: `${fixturesDirectory}/try-handle-recall/basic-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall in Try",
      //   fixture: `${fixturesDirectory}/try-handle-recall/perform-in-try.js`,
      //   // outputFixture: `${fixturesDirectory}/try-handle-recall/basic-out.js`
      // },
      //
      // {
      //   title: "Try/Handle/Recall Multiple Descendants",
      //   fixture: `${fixturesDirectory}/try-handle-recall/multiple-descendents-in.js`,
      //   outputFixture: `${fixturesDirectory}/try-handle-recall/multiple-descendents-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall Assignments",
      //   fixture: `${fixturesDirectory}/try-handle-recall/assignments-in.js`,
      //   outputFixture: `${fixturesDirectory}/try-handle-recall/assignments-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall Assignments",
      //   fixture: `${fixturesDirectory}/try-handle-recall/complex-in.js`,
      //   // outputFixture: `${fixturesDirectory}/try-handle-recall/assignments-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall Handlers as functions",
      //   fixture: `${fixturesDirectory}/try-handle-recall/separated-handlers-in.js`,
      //   outputFixture: `${fixturesDirectory}/try-handle-recall/separated-handlers-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall async handlers",
      //   fixture: `${fixturesDirectory}/try-handle-recall/async-handlers-in.js`,
      //   outputFixture: `${fixturesDirectory}/try-handle-recall/async-handlers-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall nested handlers",
      //   fixture: `${fixturesDirectory}/try-handle-recall/nested-handlers-in.js`,
      //   outputFixture: `${fixturesDirectory}/try-handle-recall/nested-handlers-out.js`
      // },
      // {
      //   title: "Try/Handle/Recall nested handlers",
      //   fixture: `${fixturesDirectory}/try-handle-recall/nested-handlers-in.js`,
      //   outputFixture: `${fixturesDirectory}/try-handle-recall/nested-handlers-out.js`
      // }
      // {
      //   title: "It Should preserve try/catch",
      //   fixture: `${fixturesDirectory}/preserve-try-catch-in.js`,
      //   outputFixture: `${fixturesDirectory}/preserve-try-catch-out.js`
      // }
    ]
  });

  it("will come back", () => {});
});
