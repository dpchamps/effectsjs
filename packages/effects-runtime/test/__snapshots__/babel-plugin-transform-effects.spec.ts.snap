// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformEffects Transform proposed effects keywords into working JS inline-arrow-function.js: inline-arrow-function.js 1`] = `

const performAThing = (type, props) => perform {type, ...props};

const recallAThing = (dataToRecall) => recall dataToRecall;

const inlineChildMethod = () => performAThing('childPerform', {data : 'performed from child'});

const root = (e, data) => {
    'use effects'
    try{
        if(typeof e === 'undefined'){
            return inlineChildMethod();
        }

        return performAThing(e, {data});

    }handle e with ({data}){
        if(data === 'functionalRecall') return recallAThing(data);

        recall data
    }handle 'childPerform' with (e){
        recall e;
    }
};

const main = async (e, data) => root(e, data);



module.exports.test = ({it, expect, code}) => {
    it('Compiles and runs an inline arrow function expression that performs', async () => {
        const identity = await main('dynamic-effect', 'hello world');
        expect(identity).toBe('hello world');
    });

    it('Compiles and runs an inline arrow function expression that recalls', async () => {
        const identity = await main('anything', 'functionalRecall');
        expect(identity).toBe('functionalRecall');
    });

    it('Compiles and runs an inline arrow function that is a child method in the virtual stack', async () => {
        const identity = await main(undefined);
        expect(identity).toEqual({data : 'performed from child' });
    })
};

      ↓ ↓ ↓ ↓ ↓ ↓

const performAThing = function*(type, props) {
  return yield performEffect({
    type,
    ...props
  });
};

const inlineChildMethod = function*() {
  return yield performAThing("childPerform", {
    data: "performed from child"
  });
};

const root = (e, data) => {
  return runProgram(
    (function*() {
      yield withHandler(
        {
          *[e]({ data }, resume) {
            const result = yield function(handler) {
              return new Promise(async (res, rej) => {
                try {
                  const recallAThing = dataToRecall =>
                    stackResume(handler, dataToRecall)
                      .then(res)
                      .catch(rej);

                  if (data === "functionalRecall") return recallAThing(data);
                  return stackResume(handler, data)
                    .then(res)
                    .catch(rej);
                } catch (handlerError) {
                  rej(handlerError);
                }
              });
            };
            return yield resume(result);
          },

          *childPerform(e, resume) {
            const result = yield function(handler) {
              return new Promise(async (res, rej) => {
                try {
                  return stackResume(handler, e)
                    .then(res)
                    .catch(rej);
                } catch (handlerError) {
                  rej(handlerError);
                }
              });
            };
            return yield resume(result);
          }
        },
        (async function*() {
          if (typeof e === "undefined") {
            return yield inlineChildMethod();
          }

          return yield performAThing(e, {
            data
          });
        })()
      );
    })()
  );
};

const main = async (e, data) => root(e, data);

module.exports.test = ({ it, expect, code }) => {
  it("Compiles and runs an inline arrow function expression that performs", async () => {
    const identity = await main("dynamic-effect", "hello world");
    expect(identity).toBe("hello world");
  });
  it("Compiles and runs an inline arrow function expression that recalls", async () => {
    const identity = await main("anything", "functionalRecall");
    expect(identity).toBe("functionalRecall");
  });
  it("Compiles and runs an inline arrow function that is a child method in the virtual stack", async () => {
    const identity = await main(undefined);
    expect(identity).toEqual({
      data: "performed from child"
    });
  });
};


`;
